File: typescript/src/streaming/StreamingService.ts
/**
 * Live stream publish/subscribe service.
 *
 * Publishers broadcast StreamAnnounce and unicast StreamSegment to each
 * subscriber. Subscribers send StreamSubscribe / StreamUnsubscribe to the
 * publisher.
 *
 * Binary segment payload:
 *   [16] StreamId (UUID RFC4122 big-endian)
 *   [4]  Sequence (uint32 little-endian)
 *   [8]  TimestampMs (int64 little-endian)
 *   [1]  IsKeyframe (0 or 1)
 *   [N]  EncodedPayload
 *
 * SPDX-License-Identifier: MIT
 */

import { DEFAULT_TTL } from "../constants.js";
import { MeshPacket } from "../protocol/MeshPacket.js";
import { PacketType } from "../protocol/PacketType.js";
import { IMeshSender } from "../routing/IMeshSender.js";
import { uuidToBytes, bytesToUuid } from "../voice/VoiceCallService.js";

// ──────────────────────────────────────────────────────────────────
// Constants
// ──────────────────────────────────────────────────────────────────

const ANNOUNCE_PRIORITY = 0;
const SEGMENT_PRIORITY = 32;
const SUB_PRIORITY = 16;

// ──────────────────────────────────────────────────────────────────
// Wire types (snake_case)
// ──────────────────────────────────────────────────────────────────

type StreamState = "live" | "ending" | "ended";

interface StreamAnnouncePayload {
  stream_id: string;
  title: string;
  content_type: string;
  codec: string;
  segment_duration_ms: number;
  state: StreamState;
  started_at_ms: number;
}

interface StreamSubscribePayload {
  stream_id: string;
  live_only: boolean;
}

interface StreamUnsubscribePayload {
  stream_id: string;
}

// ──────────────────────────────────────────────────────────────────
// Public types
// ──────────────────────────────────────────────────────────────────

export interface StreamInfo {
  streamId: string;
  publisherUhid: string;
  title: string;
  contentType: string;
  codec: string;
  segmentDurationMs: number;
  state: StreamState;
  startedAtMs: number;
}

export interface SegmentEvent {
  streamId: string;
  fromUhid: string;
  sequence: number;
  timestampMs: number;
  isKeyframe: boolean;
  encodedPayload: Uint8Array;
}

// ──────────────────────────────────────────────────────────────────
// Service
// ──────────────────────────────────────────────────────────────────

export class StreamingService {
  /** streams we are publishing: streamId → info */
  private readonly publishedStreams = new Map<string, StreamInfo>();
  /** streams we are subscribed to: streamId → publisherUhid */
  private readonly subscribedStreams = new Map<string, string>();
  /** subscribers per stream we are publishing: streamId → Set<uhid> */
  private readonly subscribers = new Map<string, Set<string>>();
  /** known remote streams: streamId → info */
  private readonly knownStreams = new Map<string, StreamInfo>();

  private segmentSeq = 0;

  // Event callbacks
  onStreamAnnounced?: (info: StreamInfo) => void;
  onSubscriberJoined?: (streamId: string, uhid: string) => void;
  onSubscriberLeft?: (streamId: string, uhid: string) => void;
  onSegmentReceived?: (event: SegmentEvent) => void;
  onStreamEnded?: (streamId: string) => void;

  constructor(private readonly sender: IMeshSender) {}

  // ──────────────── publisher actions ─────────────────────────────

  async startStream(
    title: string,
    contentType: string,
    codec: string,
    segmentDurationMs: number,
  ): Promise<string> {
    const streamId = crypto.randomUUID();
    const now = Date.now();
    const info: StreamInfo = {
      streamId,
      publisherUhid: this.sender.localUhid,
      title,
      contentType,
      codec,
      segmentDurationMs,
      state: "live",
      startedAtMs: now,
    };
    this.publishedStreams.set(streamId, info);
    this.subscribers.set(streamId, new Set<string>());

    await this.broadcastAnnounce(info);
    return streamId;
  }

  async endStream(streamId: string): Promise<void> {
    const info = this.publishedStreams.get(streamId);
    if (!info) return;

    info.state = "ending";
    await this.broadcastAnnounce(info);

    info.state = "ended";
    await this.broadcastAnnounce(info);

    this.publishedStreams.delete(streamId);
    this.subscribers.delete(streamId);
    this.onStreamEnded?.(streamId);
  }

  async publishSegment(
    streamId: string,
    data: Uint8Array,
    isKeyframe: boolean,
  ): Promise<void> {
    const info = this.publishedStreams.get(streamId);
    if (!info || info.state !== "live") return;

    const subs = this.subscribers.get(streamId);
    if (!subs || subs.size === 0) return;

    const payload = encodeStreamSegment(streamId, data, isKeyframe, this.segmentSeq++);

    for (const uhid of subs) {
      const packet = new MeshPacket();
      packet.type = PacketType.StreamSegment;
      packet.sourceUhid = this.sender.localUhid;
      packet.destinationUhid = uhid;
      packet.ttl = DEFAULT_TTL;
      packet.priority = SEGMENT_PRIORITY;
      packet.payload = payload;
      await this.sender.send(packet, uhid);
    }
  }

  // ──────────────── subscriber actions ────────────────────────────

  async subscribe(
    streamId: string,
    publisherUhid: string,
    liveOnly: boolean,
  ): Promise<void> {
    this.subscribedStreams.set(streamId, publisherUhid);

    const body = new TextEncoder().encode(
      JSON.stringify({ stream_id: streamId, live_only: liveOnly } satisfies StreamSubscribePayload),
    );
    const packet = new MeshPacket();
    packet.type = PacketType.StreamSubscribe;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = publisherUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = SUB_PRIORITY;
    packet.payload = body;
    await this.sender.send(packet, publisherUhid);
  }

  async unsubscribe(streamId: string, publisherUhid: string): Promise<void> {
    this.subscribedStreams.delete(streamId);

    const body = new TextEncoder().encode(
      JSON.stringify({ stream_id: streamId } satisfies StreamUnsubscribePayload),
    );
    const packet = new MeshPacket();
    packet.type = PacketType.StreamUnsubscribe;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = publisherUhid;
    packet.ttl = DEFAULT_TTL;
    packet.priority = SUB_PRIORITY;
    packet.payload = body;
    await this.sender.send(packet, publisherUhid);
  }

  // ──────────────── inbound packet routing ────────────────────────

  async onPacket(packet: MeshPacket): Promise<void> {
    switch (packet.type) {
      case PacketType.StreamAnnounce:
        this.handleAnnounce(packet);
        break;
      case PacketType.StreamSubscribe:
        this.handleSubscribe(packet);
        break;
      case PacketType.StreamUnsubscribe:
        this.handleUnsubscribe(packet);
        break;
      case PacketType.StreamSegment:
        this.handleSegment(packet);
        break;
      default:
        break;
    }
  }

  // ──────────────── inspection ─────────────────────────────────────

  getKnownStreams(): StreamInfo[] {
    return Array.from(this.knownStreams.values());
  }

  getSubscribers(streamId: string): string[] {
    return Array.from(this.subscribers.get(streamId) ?? []);
  }

  // ──────────────── private helpers ─────────────────────────────────

  private async broadcastAnnounce(info: StreamInfo): Promise<void> {
    const body = new TextEncoder().encode(
      JSON.stringify({
        stream_id: info.streamId,
        title: info.title,
        content_type: info.contentType,
        codec: info.codec,
        segment_duration_ms: info.segmentDurationMs,
        state: info.state,
        started_at_ms: info.startedAtMs,
      } satisfies StreamAnnouncePayload),
    );
    const packet = new MeshPacket();
    packet.type = PacketType.StreamAnnounce;
    packet.sourceUhid = this.sender.localUhid;
    packet.destinationUhid = "";
    packet.ttl = DEFAULT_TTL;
    packet.priority = ANNOUNCE_PRIORITY;
    packet.payload = body;
    await this.sender.broadcast(packet);
  }

  private handleAnnounce(packet: MeshPacket): void {
    let msg: StreamAnnouncePayload;
    try {
      msg = JSON.parse(
        new TextDecoder().decode(packet.payload),
      ) as StreamAnnouncePayload;
    } catch {
      return;
    }
    if (!msg.stream_id) return;

    const info: StreamInfo = {
      streamId: msg.stream_id,
      publisherUhid: packet.sourceUhid,
      title: msg.title ?? "",
      contentType: msg.content_type ?? "",
      codec: msg.codec ?? "",
      segmentDurationMs: msg.segment_duration_ms ?? 0,
      state: msg.state,
      startedAtMs: msg.started_at_ms ?? Date.now(),
    };

    if (msg.state === "ended") {
      this.knownStreams.delete(msg.stream_id);
      this.subscribedStreams.delete(msg.stream_id);
      this.onStreamEnded?.(msg.stream_id);
    } else {
      this.knownStreams.set(msg.stream_id, info);
      this.onStreamAnnounced?.(info);
    }
  }

  private handleSubscribe(packet: MeshPacket): void {
    let msg: StreamSubscribePayload;
    try {
      msg = JSON.parse(
        new TextDecoder().decode(packet.payload),
      ) as StreamSubscribePayload;
    } catch {
      return;
    }
    if (!msg.stream_id) return;
    if (!this.publishedStreams.has(msg.stream_id)) return;

    const subs = this.subscribers.get(msg.stream_id) ?? new Set<string>();
    subs.add(packet.sourceUhid);
    this.subscribers.set(msg.stream_id, subs);
    this.onSubscriberJoined?.(msg.stream_id, packet.sourceUhid);
  }

  private handleUnsubscribe(packet: MeshPacket): void {
    let msg: StreamUnsubscribePayload;
    try {
      msg = JSON.parse(
        new TextDecoder().decode(packet.payload),
      ) as StreamUnsubscribePayload;
    } catch {
      return;
    }
    if (!msg.stream_id) return;

    const subs = this.subscribers.get(msg.stream_id);
    if (subs?.has(packet.sourceUhid)) {
      subs.delete(packet.sourceUhid);
      this.onSubscriberLeft?.(msg.stream_id, packet.sourceUhid);
    }
  }

  private handleSegment(packet: MeshPacket): void {
    const seg = decodeStreamSegment(packet.payload);
    if (!seg) return;

    this.onSegmentReceived?.({
      streamId: seg.streamId,
      fromUhid: packet.sourceUhid,
      sequence: seg.sequence,
      timestampMs: seg.timestampMs,
      isKeyframe: seg.isKeyframe,
      encodedPayload: seg.encodedPayload,
    });
  }
}

// ──────────────────────────────────────────────────────────────────
// Binary codec
// ──────────────────────────────────────────────────────────────────

/**
 * StreamSegment binary payload:
 * [16] StreamId (UUID RFC4122 big-endian)
 * [4]  Sequence (uint32 little-endian)
 * [8]  TimestampMs (int64 little-endian)
 * [1]  IsKeyframe (0 or 1)
 * [N]  EncodedPayload
 */
export function encodeStreamSegment(
  streamId: string,
  encodedPayload: Uint8Array,
  isKeyframe: boolean,
  sequence: number,
): Uint8Array {
  const buf = new Uint8Array(16 + 4 + 8 + 1 + encodedPayload.length);
  const dv = new DataView(buf.buffer);

  uuidToBytes(streamId, buf, 0);
  dv.setUint32(16, sequence >>> 0, true);
  dv.setBigInt64(20, BigInt(Date.now()), true);
  buf[28] = isKeyframe ? 1 : 0;
  buf.set(encodedPayload, 29);

  return buf;
}

export function decodeStreamSegment(data: Uint8Array): {
  streamId: string;
  sequence: number;
  timestampMs: number;
  isKeyframe: boolean;
  encodedPayload: Uint8Array;
} | null {
  if (data.length < 29) return null;
  const dv = new DataView(data.buffer, data.byteOffset, data.byteLength);

  const streamId = bytesToUuid(data, 0);
  const sequence = dv.getUint32(16, true);
  const timestampMs = Number(dv.getBigInt64(20, true));
  const isKeyframe = data[28] !== 0;
  const encodedPayload = data.slice(29);

  return { streamId, sequence, timestampMs, isKeyframe, encodedPayload };
}

File: kotlin/src/main/kotlin/aethernet/streaming/StreamingService.kt
// SPDX-License-Identifier: MIT

package aethernet.streaming

import aethernet.AetherNetConstants
import aethernet.protocol.MeshPacket
import aethernet.protocol.PacketType
import aethernet.routing.MeshSender
import aethernet.voice.JsonReader
import aethernet.voice.jsonEscape
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

// ─────────────────────────────────────────────────────────────────────────────
// Domain types
// ─────────────────────────────────────────────────────────────────────────────

enum class StreamState { Live, Ended }

data class StreamInfo(
    val streamId: UUID,
    val publisherUhid: String,
    val title: String,
    val contentType: String,
    val codec: String,
    val segmentDurationMs: Int,
    var state: StreamState = StreamState.Live,
    val startedAtMs: Long = System.currentTimeMillis()
)

/**
 * Live-streaming service for audio/video segment distribution over the mesh.
 *
 * Wire format — StreamSegment binary payload:
 *   [16] StreamId      (UUID RFC4122 big-endian)
 *   [4]  Sequence      (UInt32 little-endian)
 *   [8]  TimestampMs   (Int64 little-endian)
 *   [1]  IsKeyframe    (0 or 1)
 *   [N]  EncodedPayload
 *
 * Signaling packets use JSON (snake_case).
 */
class StreamingService(private val sender: MeshSender) {

    /** Active streams this node is publishing. */
    private val activeStreams = ConcurrentHashMap<UUID, StreamInfo>()

    /**
     * Subscriber map: streamId → set of subscriber UHIDs.
     * Access via ConcurrentHashMap for thread safety.
     */
    private val subscribers = ConcurrentHashMap<UUID, MutableSet<String>>()

    @Volatile private var segmentSequence: Int = 0

    /** Fired when a stream announce arrives from a remote publisher. */
    var onStreamAnnounced: ((StreamInfo) -> Unit)? = null

    /** Fired when a stream ends (state → Ended). */
    var onStreamEnded: ((UUID) -> Unit)? = null

    /** Fired when a segment arrives for a subscribed stream. */
    var onSegmentReceived: ((UUID, ByteArray, Boolean) -> Unit)? = null

    // ─────────────────────────────────────────────────────────────────────
    // Publisher API
    // ─────────────────────────────────────────────────────────────────────

    /**
     * Start a new stream and announce it to the mesh.
     */
    suspend fun startStream(
        title: String,
        contentType: String,
        codec: String,
        segmentDurationMs: Int
    ): UUID {
        require(title.isNotEmpty()) { "title must not be empty" }
        val streamId = UUID.randomUUID()
        val info = StreamInfo(
            streamId = streamId,
            publisherUhid = sender.localUhid,
            title = title,
            contentType = contentType,
            codec = codec,
            segmentDurationMs = segmentDurationMs,
            state = StreamState.Live
        )
        activeStreams[streamId] = info
        subscribers[streamId] = ConcurrentHashMap.newKeySet()

        val payload = encodeStreamAnnounce(info)
        val packet = MeshPacket(
            type = PacketType.StreamAnnounce,
            sourceUhid = sender.localUhid,
            destinationUhid = "",
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 32,
            payload = payload
        )
        sender.broadcast(packet)
        return streamId
    }

    /**
     * End the stream and announce to subscribers.
     */
    suspend fun endStream(streamId: UUID) {
        val info = activeStreams.remove(streamId) ?: return
        info.state = StreamState.Ended
        subscribers.remove(streamId)

        val payload = encodeStreamAnnounce(info)
        val packet = MeshPacket(
            type = PacketType.StreamAnnounce,
            sourceUhid = sender.localUhid,
            destinationUhid = "",
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 32,
            payload = payload
        )
        sender.broadcast(packet)
        onStreamEnded?.invoke(streamId)
    }

    /**
     * Subscribe to a remote stream. Sends a subscribe request to [publisherUhid].
     */
    suspend fun subscribe(streamId: UUID, publisherUhid: String, liveOnly: Boolean) {
        require(publisherUhid.isNotEmpty()) { "publisherUhid must not be empty" }
        val payload = encodeStreamSubscribe(streamId, liveOnly)
        val packet = MeshPacket(
            type = PacketType.StreamSubscribe,
            sourceUhid = sender.localUhid,
            destinationUhid = publisherUhid,
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 32,
            payload = payload
        )
        sender.send(packet, publisherUhid)
    }

    /**
     * Unsubscribe from a remote stream.
     */
    suspend fun unsubscribe(streamId: UUID, publisherUhid: String) {
        require(publisherUhid.isNotEmpty()) { "publisherUhid must not be empty" }
        val payload = encodeStreamUnsubscribe(streamId)
        val packet = MeshPacket(
            type = PacketType.StreamUnsubscribe,
            sourceUhid = sender.localUhid,
            destinationUhid = publisherUhid,
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 32,
            payload = payload
        )
        sender.send(packet, publisherUhid)
    }

    /**
     * Publish a new segment to all current subscribers of [streamId].
     */
    suspend fun publishSegment(streamId: UUID, data: ByteArray, isKeyframe: Boolean) {
        if (!activeStreams.containsKey(streamId)) return
        val subs = subscribers[streamId] ?: return

        val seq = segmentSequence++.toUInt()
        val payload = encodeStreamSegment(streamId, seq, System.currentTimeMillis(), isKeyframe, data)
        val packet = MeshPacket(
            type = PacketType.StreamSegment,
            sourceUhid = sender.localUhid,
            destinationUhid = "",
            ttl = AetherNetConstants.DEFAULT_TTL,
            priority = 48,
            payload = payload
        )
        for (sub in subs) {
            sender.send(packet.copy(destinationUhid = sub), sub)
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Inbound packet dispatcher
    // ─────────────────────────────────────────────────────────────────────

    suspend fun handlePacket(packet: MeshPacket) {
        when (packet.type) {
            PacketType.StreamAnnounce -> handleAnnounce(packet)
            PacketType.StreamSubscribe -> handleSubscribe(packet)
            PacketType.StreamUnsubscribe -> handleUnsubscribe(packet)
            PacketType.StreamSegment -> handleSegment(packet)
            else -> {}
        }
    }

    // ─────────────────────────────────────────────────────────────────────
    // Private helpers
    // ─────────────────────────────────────────────────────────────────────

    private fun handleAnnounce(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val streamIdStr = JsonReader.readString(json, "stream_id") ?: return
        val streamId = runCatching { UUID.fromString(streamIdStr) }.getOrNull() ?: return
        val title = JsonReader.readString(json, "title") ?: ""
        val contentType = JsonReader.readString(json, "content_type") ?: ""
        val codec = JsonReader.readString(json, "codec") ?: ""
        val segmentMs = JsonReader.readInt(json, "segment_duration_ms") ?: AetherNetConstants.DEFAULT_SEGMENT_DURATION_MS.toInt()
        val stateStr = JsonReader.readString(json, "state") ?: "live"
        val startedAtMs = JsonReader.readLong(json, "started_at_ms") ?: System.currentTimeMillis()
        val state = if (stateStr == "ended") StreamState.Ended else StreamState.Live

        val info = StreamInfo(
            streamId = streamId,
            publisherUhid = packet.sourceUhid,
            title = title,
            contentType = contentType,
            codec = codec,
            segmentDurationMs = segmentMs,
            state = state,
            startedAtMs = startedAtMs
        )
        if (state == StreamState.Ended) {
            onStreamEnded?.invoke(streamId)
        } else {
            onStreamAnnounced?.invoke(info)
        }
    }

    private fun handleSubscribe(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val streamIdStr = JsonReader.readString(json, "stream_id") ?: return
        val streamId = runCatching { UUID.fromString(streamIdStr) }.getOrNull() ?: return
        if (!activeStreams.containsKey(streamId)) return

        subscribers.getOrPut(streamId) { ConcurrentHashMap.newKeySet() }
            .add(packet.sourceUhid)
    }

    private fun handleUnsubscribe(packet: MeshPacket) {
        val json = String(packet.payload, Charsets.UTF_8)
        val streamIdStr = JsonReader.readString(json, "stream_id") ?: return
        val streamId = runCatching { UUID.fromString(streamIdStr) }.getOrNull() ?: return
        subscribers[streamId]?.remove(packet.sourceUhid)
    }

    private fun handleSegment(packet: MeshPacket) {
        val data = packet.payload
        // Minimum: 16 + 4 + 8 + 1 = 29 bytes
        if (data.size < 29) return

        val uuidBuf = ByteBuffer.wrap(data, 0, 16).order(ByteOrder.BIG_ENDIAN)
        val streamId = UUID(uuidBuf.long, uuidBuf.long)

        val buf = ByteBuffer.wrap(data).order(ByteOrder.LITTLE_ENDIAN)
        buf.position(16)
        @Suppress("UNUSED_VARIABLE") val seq = buf.int.toUInt()
        @Suppress("UNUSED_VARIABLE") val ts = buf.long
        val isKeyframe = buf.get() != 0.toByte()
        val encoded = if (buf.hasRemaining()) {
            val arr = ByteArray(buf.remaining())
            buf.get(arr)
            arr
        } else ByteArray(0)

        onSegmentReceived?.invoke(streamId, encoded, isKeyframe)
    }
}

// ─────────────────────────────────────────────────────────────────────────────
// Wire helpers
// ─────────────────────────────────────────────────────────────────────────────

internal fun encodeStreamSegment(
    streamId: UUID,
    sequence: UInt,
    timestampMs: Long,
    isKeyframe: Boolean,
    encodedPayload: ByteArray
): ByteArray {
    val buf = ByteBuffer.allocate(16 + 4 + 8 + 1 + encodedPayload.size)
    buf.order(ByteOrder.BIG_ENDIAN)
    buf.putLong(streamId.mostSignificantBits)
    buf.putLong(streamId.leastSignificantBits)
    buf.order(ByteOrder.LITTLE_ENDIAN)
    buf.putInt(sequence.toInt())
    buf.putLong(timestampMs)
    buf.put(if (isKeyframe) 1.toByte() else 0.toByte())
    buf.put(encodedPayload)
    return buf.array()
}

private fun encodeStreamAnnounce(info: StreamInfo): ByteArray {
    val sb = StringBuilder()
    sb.append('{')
    sb.append("\"stream_id\":\"").append(info.streamId).append("\",")
    sb.append("\"title\":\"").append(jsonEscape(info.title)).append("\",")
    sb.append("\"content_type\":\"").append(jsonEscape(info.contentType)).append("\",")
    sb.append("\"codec\":\"").append(jsonEscape(info.codec)).append("\",")
    sb.append("\"segment_duration_ms\":").append(info.segmentDurationMs).append(',')
    sb.append("\"state\":\"").append(if (info.state == StreamState.Live) "live" else "ended").append("\",")
    sb.append("\"started_at_ms\":").append(info.startedAtMs)
    sb.append('}')
    return sb.toString().toByteArray(Charsets.UTF_8)
}

private fun encodeStreamSubscribe(streamId: UUID, liveOnly: Boolean): ByteArray {
    val s = "{\"stream_id\":\"$streamId\",\"live_only\":$liveOnly}"
    return s.toByteArray(Charsets.UTF_8)
}

private fun encodeStreamUnsubscribe(streamId: UUID): ByteArray {
    val s = "{\"stream_id\":\"$streamId\"}"
    return s.toByteArray(Charsets.UTF_8)
}
